torch/testing/_internal/common_utils.py

```diff
@@ -803,204 +810,412 @@ def is_integral(self, dtype):
         dtypes = [x for x in get_all_dtypes() if x not in get_all_complex_dtypes()]
         return dtype in dtypes and not dtype.is_floating_point
 
-    # accepts tensors, dtypes, or np.ndarrays
-    def get_default_tolerance(self, a, b=None):
-        if b is None:
-            dtype = torch.float
-            if isinstance(a, torch.Tensor):
-                dtype = a.dtype
-            elif isinstance(a, torch.dtype):
-                dtype = a
-            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
-                # Some tests call assertEqual with numpy Unicode.
-                if numpy.issubdtype(a.dtype, numpy.dtype('U')):
-                    dtype = torch.float
-                else:
-                    dtype = torch.from_numpy(a).dtype
-            if self.is_integral(dtype):
-                return (0, 0)
-            dtype = str(dtype).split('.')[-1]
-            return self.dtype_precisions.get(dtype, (self.precision, self.precision))
-
-        a_tol = self.get_default_tolerance(a)
-        b_tol = self.get_default_tolerance(b)
-        return (max(a_tol[0], b_tol[0]), max(a_tol[1], b_tol[1]))
-
-    def assertEqual(self, x, y, message='', *, atol=None, rtol=None, allow_inf=False, exact_dtype=None):
+    # Returns a dtype suitable for comparing tensors a and b.
+    # This closely follows NumPy's type promotion logic, except it includes
+    # bfloat16 and accounts for the current state of PyTorch's bfloat16 and
+    # float16 implementations.
+    # Note: since torch.isclose converts its inputs to floating point values
+    # (it multiplies by rtol, a double) this function takes an optional
+    # allow_ints argument. If false, it only returns bool, float, or complex
+    # tensors. This can prevent a second conversion when comparing tensors.
+    def compare_type(self, a, b, *, allow_ints=True):
+
+        class _TypeMeta:
+            __slots__ = ['rank', 'kind', 'width']
+
+            def __init__(self, rank, kind, width):
+                self.rank = rank
+                self.kind = kind
+                self.width = width
+
+
+        # Maps torch dtypes to their rank, kind and width
+        type_meta = {
+            torch.complex128 : _TypeMeta(11, 'complex', 64),
+            torch.complex64  : _TypeMeta(10, 'complex', 32),
+            torch.float64    : _TypeMeta(9, 'float', 64),
+            torch.float32    : _TypeMeta(8, 'float', 32),
+            torch.bfloat16   : _TypeMeta(7, 'float', 16),
+            torch.float16    : _TypeMeta(6, 'float', 16),
+            torch.int64      : _TypeMeta(5, 'int', 64),
+            torch.int32      : _TypeMeta(4, 'int', 32),
+            torch.int16      : _TypeMeta(3, 'int', 16),
+            torch.int8       : _TypeMeta(2, 'int', 8),
+            torch.uint8      : _TypeMeta(1, 'uint', 8),
+            torch.bool       : _TypeMeta(0, 'uint', 1),
```

Comment:
Is the new "rank" concept necessary? I'm afraid it makes it harder to read and understand. We already have "category" and an existing type promotion system, and below we just seem to get a maximum or to check for specific dtypes.

Do we need quantized types here?

Reply:
We don't need quantized types (today, anyway) because assertEqual casts the integer representation of quantized types to int32 before comparing them. 

The two best ways to simplify this code would be to take a NumPy dependency and use NumPy's type promotion or not care about trying to preserve int x float comparison numerics and use PyTorch's type promotion.

Reply:
It's pretty likely that someone will want to add a uint64 and/or uint128 at some point for cryptography code. To do so they'll need to be sure to adjust the 7 and 8 literals below. I think it'll be a bit awkward for them when reading this what we're ranking them by. e.g. what makes bfloat16 higher `rank` than float16? Do we rely on the fact that float32 is higher rank than int64, or is the rank only actually meaningful within a "kind"?

Reply:
FYI, we have functions like finfo and iinfo that could help here:
```
>>> torch.finfo(torch.float).bits
32
>>> torch.iinfo(torch.int).bits
32
```
We may want to introducing additional type metadata concepts soon as well for a subdtype API.

Reply:
I spent a considerable amount of time looking at alternative implementations for this function, and unfortunately I think NumPy's type promotion is sufficiently complicated to justify the current approach. The rank is very useful for simplifying the code, and by the time we implement uint128 we'll probably have a NumPy dependency that simplifies the code or, yes, someone will have to implement the appropriate promotion for it.

We could use torch.finfo and torch.iinfo instead of inline some of the constants, but torch.bool and the complex types would still require custom values. I'm also not sure how advantageous it is to acquire those values vs. just inlining the constants?

Reply:
Per later comment: I will replace this function with a table.